src/discordtipbot/cache.cr

```diff
@@ -0,0 +1,84 @@
+class Cache
+  # Container for a snowflake and the last time it was seen
+  record Entry, id : UInt64, seen : Time do
+    def ==(other)
+      id == other.id
+    end
+
+    def <(other)
+      seen < other.seen
+    end
+
+    def >(other)
+      seen > other.seen
+    end
+
+    def hash
+      id.hash
+    end
+
+    def older_than?(age : Time::Span)
+      Time.now > seen + age
+    end
+  end
+
+  @cache = {} of UInt64 => Set(Entry)
+
+  def initialize(@ttl : Time::Span)
+  end
+
+  # Marks the `user_id` as being observed in `channel_id` at the given `time`
+  def touch(channel_id : UInt64, user_id : UInt64, time = Time.now)
+    entry = Entry.new(user_id, time)
+    set = @cache[channel_id] ||= Set(Entry).new
+    set.delete(entry) if set.includes?(entry)
+    set.add(entry)
+    @cache[channel_id] = set
+  end
+
+  # Only marks `user_id` as being observed in `channel_id` if `time` is newer then latest entry
+  def add_if_youngest(channel_id : UInt64, user_id : UInt64, time : Time)
```

Comment:
No, if you want this behavior, put it into `#touch` so it simply skips `set.add` if the entry is already `#older_than?(@ttl)`. Or just do not `#touch` users into the cache when parsing message history if its older than TTL.

Reply:
Hm. I think I'll just go the older_than route in that case